### PR TITLE
[addons][gui] Add method to get allowed/whitelist resolutions

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

Add a method to get allowed and whitelist resolutions from C++ addon interface

## Motivation and context

On InputStream Adaptive we need to know the maximum allowed achievable resolution
so as to choose the appropriate video stream resolution to not be bound to the GUI resolution

i need a little help to finish it

## How has this been tested?

manually modified/build InputStream Adaptive by adding this code somewhere:

```c++
#include <kodi/gui/General.h>

...

auto list = kodi::gui::GetWhitelistResolutions();
for (auto& res : list)
{
      LOG::Log(LOGDEBUG, "WHITELIST AVAILABLE %ix%i - %f", res.GetWidth(), res.GetHeight(), res.GetRefreshRate());
}
```

## What is the effect on users?

Not directly, this allow us to improve ISAdaptive core

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h [hunk 1]
```diff
@@ -170,6 +191,107 @@ inline kodi::HardwareContext GetHWContext()
...
+/// ~~~~~~~~~~~~~
+///
+inline std::vector<SResolutionEntry> GetWhitelistResolutions()
+{
+  using namespace ::kodi::addon;
+  std::vector<SResolutionEntry> resList;
+  int* widthList{nullptr};
+  int* heightList{nullptr};
+  float* refreshRateList{nullptr};
+  unsigned int size{0};
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h:217]: About the planned sandbox is would always better to use for any size values a `size_t` as easier to identify from scripts.

File: xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h [hunk 2]
```diff
@@ -170,6 +191,107 @@ inline kodi::HardwareContext GetHWContext()
...
+
+  CPrivateBase::m_interface->toKodi->kodi_gui->general->get_whitelist_resolutions(
+      CPrivateBase::m_interface->toKodi->kodiBase, &widthList, &heightList, &refreshRateList, &size);
+
+  if (widthList && heightList && refreshRateList)
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h:222]: With structure only one check needed:
```cpp
inline std::vector<SResolutionEntry> GetWhitelistResolutions()
{
  using namespace ::kodi::addon;
  std::vector<SResolutionEntry> resList;

  KODI_GUI_RESOLUTION_INFO* infoList{nullptr};
  size_t size{0};
  CPrivateBase::m_interface->toKodi->kodi_gui->general->get_whitelist_resolutions(
      CPrivateBase::m_interface->toKodi->kodiBase, &infoList, &size);

  if (infoList)
  {
    for (size_t i{0}; i < size; i++)
    {
      resList.emplace_back(infoList[i]);
    }

    free(infoList);
  }

File: xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h [hunk 3]
```diff
@@ -170,6 +191,107 @@ inline kodi::HardwareContext GetHWContext()
...
+      entry.height = heightList[i];
+      entry.refreshRate = refreshRateList[i];
+      resList.emplace_back(entry);
+    }
+  }
+
+  /* Debug assertion failed _crtisValidHeapPointer(block)
+  free(widthList);
+  free(heightList);
+  free(refreshRateList);
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h:237]: Better to call inside `if` place.

File: xbmc/addons/interfaces/gui/General.cpp [hunk 4]
```diff
@@ -212,6 +215,66 @@ ADDON_HARDWARE_CONTEXT Interface_GUIGeneral::get_hw_context(KODI_HANDLE kodiBase
   return CServiceBroker::GetWinSystem()->GetHWContext();
 }
 
+void Interface_GUIGeneral::get_whitelist_resolutions(KODI_HANDLE kodiBase,
+                                                     int** widthList,
+                                                     int** heightList,
+                                                     float** refreshRateList,
+                                                     unsigned int* size)
+{
```
[the remark this answers is not in the document]
Comment (CastagnaIT) [reply xbmc/addons/interfaces/gui/General.cpp:221]: done,
i found that `Debug assertion failed _crtisValidHeapPointer(block)` error on `free` call is raised because i build ISAdaptive as debug and Kodi as release
for some reason is influenced by this

File: xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h [hunk 5]
```diff
@@ -17,6 +17,27 @@ namespace kodi
...
+
+  /// Width
+  int width;
+
+  /// Height
+  int height;
+
+  /// Refresh rate
+  float refreshRate;
+} SResolutionEntry;
```
[the remark this answers is not in the document]
Comment (CastagnaIT) [reply xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h]: done

File: xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h (hunk 1 above)
Comment (CastagnaIT) [reply xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h]: done

File: xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h (hunk 2 above)
Comment (CastagnaIT) [reply xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h]: done

File: xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h (hunk 3 above)
Comment (CastagnaIT) [reply xbmc/addons/kodi-dev-kit/include/kodi/gui/General.h]: done

**Comment by lrusak** [comment]
I'm not really convinced that this is a good idea. Can you explain a bit more why this change is needed?

We don't have any other bounding of the video playback to the allowed resolution, we just handle whatever is chosen.

Can't you just have a limit Max resolution option in inputstream (doesn't this already exist)?

**Comment by CastagnaIT** [comment]
no the problem is another,
is mostly intended for kodi setting "Adjust refresh rate" to start/stop
this feature change the screen resolution to match the video stream resolution

currently with ISAdaptive we dont know what is the max resolution allowed,
we know only the current GUI resolution that may not match the max resolution supported/allowed

Most use cases:
- 4k tv with android, have the kodi gui at 1080p, but can reach 4k resolution at playback start
- GUI 1080p but an user limit max resolution by set whitelist res to max 720p (cases of RPI's are slower CPU's that are not able to decode HD DRM streams, limits the res. in this way, not so right, but us having an adaptive stream implementation can be fine, because it is part of the behaviour to be followed for adaptive streams)

then when we process a stream manifest, we are not able to determine which stream resolution to use to start the playback
so we are limited to follow the GUI resolution (in the example above 1080p)
where instead we should send a different resolution, for example to allow "Adjust refresh rate" to works properly

until now the only way that we have found, is to add a ISA workaroud setting "ignore display resolution", that cover at least the 4k case, so this fake the maximum resolution value by ignoring the GUI resolution value,
therefore we instruct kodi to start the playback with the stream with the max resolution, with the side effect that may exceed the kodi res. allowed and also not optimise network bandwidth

every user therefore has to be informed to enable this workaround or will have resolution problems, it is inconvenient,
since we can get the resolutions we can remove this workaround

**Comment by CastagnaIT** [comment]
adaptive stream technology need to know in full device capabilities and device informations
this to optimize the video experience, that concern device resolutions, cpu busy level, bandwidth, drm characteristics ecc...
if you don't believe what I'm trying to make you understand, do a research on how works adaptive stream technology

if i dont know the device capability the adaptive stream cannot work properly
and each user will have to play with workarounds settings in ISAdaptive which makes no sense whatsoever...
it should not be a task of each user to worry about set the workaround settings before start video playback because we have a faulty adaptive implementation

I think i have provided enough information on the problem
we are not talking about physical files where the resolution is fixed and you don't care what the device supports,
streaming is another world and to optimize it we need all device informations
I am tring to make things work as adaptive stream technology is intended

**Comment by CastagnaIT** [comment]
if you don't like provide all resolutions by interface i need at least expose two methods
 - GetMaxSupportedResolution
 - GetMaxAllowedResolution

**Comment by lrusak** [comment]
> - GUI 1080p but an user limit max resolution by set whitelist res to max 720p (cases of RPI's are slower CPU's that are not able to decode HD DRM streams, limits the res. in this way, not so right, but us having an adaptive stream implementation can be fine, because it is part of the behaviour to be followed for adaptive streams)
> 

This is an abuse of the whitelist and shouldn't be considered for a design decision for inputstream resolution selection.

Again, you could just define a maximum resolution in inputstream that the user can set. I don't see how knowing all the available resolutions help here. On RPi you will always have to manually decide what the software decoder is capable of. There is no one size fits all here.

Streaming can have resolution changes built in (so can fixed files btw) but this shouldn't cause a resolution change in kodi AFAIK. I wouldn't want my tv to be changing resolution all the time because a stream/file has a bunch of resolution changes in it. This should be handled in the video renderer through scaling. Optimizing this is basically impossible because you won't know what's available.

Many people don't care about resolution switching as most services/applications don't even support it. We have always tried to provide the simplest behaviour for resolution switching for those that want it and of course there will always be edge cases. This is why we created the whitelist. To allow further flexibility for the user to decide what they want without having to write in every edge case. It's to be used by those that need to fine tune the behaviour of Kodi to fit their needs. One shouldn't be using the whitelist if they don't have a valid need.

So interms of this PR I don't see why you need to know the exact resolutions available to select a valid stream.

I would rather see the option of videoplayer responding back to inputstream in the case of an invalid configuration. This could then be extended beyond just resolutions. I would rather not just keep extending the add-on API for edge case usage.

This of course is just my 2c and perhaps someone else can jump in with their opinion about inputstream/videoplayer.

**Comment by CastagnaIT** [comment]
nooo :( you are confusing display resolution change, with, streaming resolution change!
only this last one takes care adaptive streaming technology

the display resolution change can takes place only when playback is started and only if "Adjust refresh rate" is enabled
(and this is one of the reasons why we need to know the maximum capable resolution)

for information only, adaptive streaming technology today is implemented in all streaming video services (internet and tv boxes) on all app's and website's, so is the basic focus of video streaming
from what I know kodi itself do not support it and actually there is no inputStream addons that have adaptive streaming technology, except for the first WIP implementation in ISAdaptive for Nexus

as I said before, I don't need to have the full list
and i have just thought of a possible third solution that promises to be better

in short i will change this method:
https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/Inputstream.h#L1563
from:
`void SetVideoResolution(int width, int height) override;`

in to:
`void SetVideoResolution(int width, int height, int widthMax, int heightMax) override;`

to provide the max res. allowed

as soon as I have time will do this test, and if works i will push a new PR to replace this one

**Comment by CastagnaIT** [comment]
@lrusak opened the alternative way #21319
i hope this time i wrote a better explanation in PR description
